src/rust/lib/data/src/hash_map_tree.rs

```diff
@@ -0,0 +1,111 @@
+//! A tree structure build on top of the `HashMap`.
+
+use crate::prelude::*;
+
+
+
+// ===================
+// === HashMapTree ===
+// ===================
+
+/// A tree build on top of the `HashMap`. Each node in the tree can have zero or more branches
+/// accessible by the given key type.
+#[derive(Derivative)]
+#[derivative(Debug   (bound="K:Eq+Hash+Debug , T:Debug"))]
+#[derivative(Default (bound="K:Eq+Hash       , T:Default"))]
+pub struct HashMapTree<K,T> {
+    /// Value of the current tree node.
+    pub value : T,
+    /// Branches of the current tree node.
+    pub branches : HashMap<K,HashMapTree<K,T>>
+}
+
+impl<K,T> HashMapTree<K,T>
+where K:Eq+Hash {
+    /// Constructor.
+    pub fn new() -> Self where T:Default {
+        default()
+    }
+
+    /// Constructor with explicit root value.
+    pub fn from_value(value:T) -> Self {
+        let branches = default();
+        Self {value,branches}
+    }
+
+    /// Iterates over keys in `path`. For each key, traverses into the appropriate branch. In case
+    /// the branch does not exist, a default instance will be created.
+    #[inline]
```

Comment:
Why we use `#[inline]` here? Usually in c++ manually inlining is disregarded, because if user want to have a fast application he should use lto, and lto usually optimizes much better that we could do.

Reply:
Stdlib uses it here as well. As agreed via chat, lets leave it.